### ISSUE TITLE
Show / Hide / Enable / Disable widgets

### ISSUE THREAD

**Issue by redtide** [cycfi/elements#125]
I haven't found any function that can make some element to be visible or not, neither to enable or disable, is there some way currently?

**Comment by Xeverous** [comment]
`popup`? If not, you can always make your own by inheriting from the desired element and overriding `draw`. If you want the disable element to also not occupy space, override `limits` aswell.

**Comment by djowel** [comment]
> I haven't found any function that can make some element to be visible or not, neither to enable or disable, is there some way currently?

Good point. It's very easy to add a proxy class that allows it to inhibit the drawing of its subject. That is the way to do it. And yes, it should be part of the library. The visibility state should probably be a callback function like the one in `basic_menu_item_element`:

https://github.com/cycfi/elements/blob/master/lib/include/elements/element/menu.hpp#L113

**Comment by djowel** [comment]
Maybe something like this usage:

```c++
auto e = share(hidable(some_element));
e-> is_visible = []{ return show_my_element_function(); /*return show-or-hide-state*/ }
```

**Comment by djowel** [comment]
> If you want the disable element to also not occupy space, override `limits` aswell.

Perhaps we also want a `collapsible` proxy for that.

**Comment by redtide** [comment]
Thank you for the quick replies and suggestions! I'll try them soon. Do I keep this issue open for a future implementation in the library?

**Comment by Xeverous** [comment]
> Do I keep this issue open for a future implementation in the library?

I would keep it. So far all of the issues were only closed when the topic was fully exhausted (by a PR or detailed "wontfix" explanation).

**Comment by djowel** [comment]
> Thank you for the quick replies and suggestions! I'll try them soon. Do I keep this issue open for a future implementation in the library?

Yes, please keep it open. These are very easy to add. I'll see if I can add them as soon as I can.

**Comment by djowel** [comment]
I got bored... I added `hidable`:

https://github.com/cycfi/elements/blob/develop/lib/include/elements/element/misc.hpp#L216-L246

and `hcollapsible` and `vcollapsible`:

https://github.com/cycfi/elements/blob/develop/lib/include/elements/element/size.hpp#L655-L725

More on ... "enable"-ability later... It's a bit more involved.

**Comment by djowel** [comment]
> My two questions have opposite meaning so I find your answer to be ambiguous.

Simply: you can have it either way. Swapping empty box is of course another way.

**Comment by Xeverous** [comment]
> Swapping empty box is of course another way.

Actually the `deck` is what I want. I want the swapped element to occupy the same space and replace it with something different depending on some runtime condition. `deck::select` is what I actually need.

**Comment by jpcima** [comment]
Thanks I would like to use `vcollapsible`, but the component draws even when collapsed.
Should not `draw` be overriden also ?

**Comment by Xeverous** [comment]
What is the difference between hidden and collapsed elements?

- hidden - occupies space but is not beign drawn?
- collapsed - does not occupy any space, as if it did not exist?

Am I right?

**Comment by jpcima** [comment]
@Xeverous This is how I understand it also.

As illustration of the problem, there is a collapsed button.
You can see the slider thumb will overlap it.

![Capture du 2020-06-14 18-59-02](https://user-images.githubusercontent.com/17614485/84599360-b33edb80-ae71-11ea-9bba-31af41cebd50.png)

**Comment by djowel** [comment]
> Thanks I would like to use `vcollapsible`, but the component draws even when collapsed.
> Should not `draw` be overriden also ?

I would imagine that the element will not draw if it has zero space, so  there must be a bug somewhere. Admittedly, the collapsible thingies haven't been tested yet. Can you make an MCVE that I can use to try this? I am guessing that the minimum limits of the element being collapsed (a button?) is being violated, so it's not as simple as I originally thought (perhaps the collapsible subject should be swapped out an empty element when collapsed).

**Comment by jpcima** [comment]
Based on the example `SpriteSlidersAndKnobs`
https://github.com/sfztools/elements/commit/13bd7d049027b9660d0c96fd0c8d1fdf6b8cafe9

The issue is present here although not as pronounced, and I don't find how to make it so as much as illustrated above.

It has a small observable amount of overlap between button and thumb.
Enabling the `#if 0` part to disable the use collapsible, it expands itself in a different layout which is more compressed and not overlapped.

**Comment by djowel** [comment]
I get it. Yes, my intuition was correct. The initial implementation was naive and incorrect. I should swap out the subject with an empty element when collapsed. A deck would probably be sufficient. It shouldn't be difficult, but I am swamped right now. Can you wait a little bit for a few days?

**Comment by jpcima** [comment]
Sure it can wait no problem. There's no fire to the lake :)
